### PR TITLE
[Reader Vec] Reader vec in duplicate key

### PR DESCRIPTION
**Pull request by HappenLee**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/volap_scan_node.h [hunk 1]
```diff
@@ -28,24 +28,26 @@ namespace vectorized {
...
         return Status::NotSupported("Not Implemented VOlapScanNode Node::get_next scalar");
     }
-    virtual Status get_next(RuntimeState* state, Block* block, bool* eos);
-    virtual Status add_one_block(Block* block);
-    virtual Status start_scan_thread(RuntimeState* state);
-    virtual Status close(RuntimeState* state);
-
-    friend class VOlapScanner;
+    Status get_next(RuntimeState* state, Block* block, bool* eos) override;
+    Status add_blocks(std::vector<Block*>& block);
```
Comment (zuochunwei) [inline be/src/vec/exec/volap_scan_node.h:42]: i suggests you distinguish public interface from private implements, make the number of interface min

Comment (HappenLee) [reply be/src/vec/exec/volap_scan_node.h:42]: ok

File: be/src/vec/exec/volap_scan_node.cpp [hunk 2]
```diff
@@ -207,46 +162,37 @@ void VOlapScanNode::scanner_thread(VOlapScanner* scanner) {
     // scan, if this exceed threshold, we yield this thread.
     int64_t raw_rows_read = scanner->raw_rows_read();
     int64_t raw_rows_threshold = raw_rows_read + config::doris_scanner_row_num;
-    while (!eos && raw_rows_read < raw_rows_threshold) {
+    bool not_miss_block = true;
```
Comment (zuochunwei) [inline be/src/vec/exec/volap_scan_node.cpp:165]: i donot know what's the meaning of not_miss_block, rethink a more meanful name?

File: be/src/vec/exec/volap_scan_node.cpp [hunk 3]
```diff
@@ -63,67 +61,29 @@ void VOlapScanNode::transfer_thread(RuntimeState* state) {
...
-    // TODO(zc): use memory limit
-    int64_t mem_consume = __sync_fetch_and_add(&_buffered_bytes, 0);
-    if (state->fragment_mem_tracker() != nullptr) {
-        mem_limit = state->fragment_mem_tracker()->limit();
-        mem_consume = state->fragment_mem_tracker()->consumption();
-    }
-    int max_thread = _max_materialized_blocks;
-    if (config::doris_scanner_row_num > state->batch_size()) {
-        max_thread /= config::doris_scanner_row_num / state->batch_size();
+    auto block_per_scanner = config::doris_scanner_row_num / state->batch_size() +
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exec/volap_scan_node.cpp:64]: ok

File: be/src/olap/row_block2.cpp [hunk 4]
```diff
@@ -94,114 +94,153 @@ Status RowBlockV2::convert_to_row_block(RowCursor* helper, RowBlock* dst) {
...
+                    origin_column.get());
+        auto& null_map = nullable_column->get_null_map_data();
+        column = nullable_column->get_nested_column_ptr().get();
+        for (uint16_t i = 0; i < _selected_size; ++i) {
+            uint16_t row_idx = _selection_vector[i];
+            null_map.push_back(_column_vector_batches[cid]->is_null_at(row_idx));
+            nullable_mark_array[i] = null_map.back();
+        }
+    } else {
+        memset(nullable_mark_array, false, _selected_size * sizeof(bool));
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/row_block2.cpp:113]: 这个sizeof尽量是在编译期生效的，这里先保持这个代码逻辑

File: be/src/olap/row_block2.cpp [hunk 5]
```diff
@@ -94,114 +94,153 @@ Status RowBlockV2::convert_to_row_block(RowCursor* helper, RowBlock* dst) {
...
-                    column.get());
+            auto column_string = assert_cast<vectorized::ColumnString*>(column);
 
             for (uint16_t j = 0; j < _selected_size; ++j) {
-                uint16_t row_idx = _selection_vector[j];
-                auto slice = reinterpret_cast<const Slice *>(column_block(cid).cell_ptr(row_idx));
-                column_string->insert_data(slice->data, slice->size);
+                if (!nullable_mark_array[j]) {
+                    uint16_t row_idx = _selection_vector[j];
+                    auto slice = reinterpret_cast<const Slice *>(column_block(cid).cell_ptr(row_idx));
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/olap/row_block2.cpp:137]: 这部分是一段临时使用的代码，保证逻辑没问题就行了。后续王博合入segment 向量化之后，这部分 代码可以剔除

File: be/src/vec/exec/volap_scan_node.cpp [hunk 6]
```diff
@@ -63,67 +61,29 @@ void VOlapScanNode::transfer_thread(RuntimeState* state) {
...
+            static_cast<int>(config::doris_scanner_row_num % state->batch_size() != 0);
+
+    for (int i = 0; i < _volap_scanners.size() * block_per_scanner; ++i) {
+        auto block = new Block;
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exec/volap_scan_node.cpp:68]: ok

File: be/src/vec/exec/volap_scan_node.cpp (hunk 2 above)
Comment (HappenLee) [reply be/src/vec/exec/volap_scan_node.cpp:165]: ok

File: be/src/vec/olap/block_reader.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    auto res = _collect_iter->next(block);
+    if (UNLIKELY(res != OLAP_SUCCESS && res != OLAP_ERR_DATA_EOF)) {
+        return res;
+    }
+    *eof = res == OLAP_ERR_DATA_EOF;
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus BlockReader::_direct_agg_key_next_block(Block* block, MemPool* mem_pool,
+                                            ObjectPool* agg_pool, bool* eof) {
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/olap/block_reader.cpp:113]: 这个是还没填充的实现，所以先简单返回了。实际要实现的代码很多，而且这个函数是作为函数指针调用的，没法内联

[the remark this answers is not in the document]
Comment (zuochunwei) [reply be/src/vec/olap/block_reader.cpp:113]: ok

**Pull request merged**